### ISSUE TITLE
cheerio's data model inconsistent with itself

### ISSUE THREAD

**Issue by rdm** [cheeriojs/cheerio#1241]
When passing data between two apis, I have html fragments from one which use protocol relative urls but the other requires absolute urls. This seems like a good reason to use cheerio. Unfortunately there seem to be a variety of problems in this context:

With cheerio version 1.0.0-rc.2,

    const cheerio= require('cheerio');
    const test= "<b><a href=\"//example.com/c\">Click here</a><br /><a href=\"//other.com/d\">Click 
    here</a></b>";
    const $= cheerio.load(test);
    for (let el= $('a[href]'); el.length; el= el.next('[href]')) {
    	const href= el.attr('href');
    	console.log(el.length, el.html(), href);
    	if (href.startsWith('//')) {
    		console.log(`fixing ${href}`);
    		el.attr('href', 'https:'+href);
    	} else {
    		console.log(`this looks ok: ${href}`);
    	}
    }
    console.log($('body').html());`

generates the result:

    fixing //example.com/c
    <b><a href="https://example.com/c">Click here</a><br><a href="https://example.com/c">Click here</a></b>

In other words, .attr('href') returns a single value but .attr('href', '...') updates two different values. This alone is a significant issue for me. I would want either .attr('href') to return two values (which I think is how jquery does it) or .attr('href', '...') to only update a single value. But the interface needs to be consistent about the data model it's representing.

This is no longer an issue for me, at this time. I can't use cheerio for my project. (I would consider trying to find the cause of these issues, and fixing them, but it's not clear from the documentation what would be acceptable as fixes). But hopefully this perspective will be of use to someone else, at some point... 

That said, note also that .next('[href]') does not find the second element with an href attribute. (But figuring out which way is up on that issue might depend how you want traversal to work for selectors.)

**Comment by meisl** [comment]
>
> In other words, .attr('href') returns a single value but .attr('href', '...') updates two different values.
> This alone is a significant issue for me.
> I would want either .attr('href') to return two values (which I think is how jquery does it) or [...]

You are simply wrong about [how jquery's .attr works](https://api.jquery.com/attr). It - and hence cheerio - does what you describe in the first sentence.
Of course one can argue about the jquery paradigm, but it won't change, that's for sure.

Here's how you'd do what you intended (slightly extended test data):
```javascript
const test = '<b>\n'
	+ '<a href="//example.com/c">Click</a><br />\n'
	+ '<a href="dont.fix.me/x">Click</a><br />\n'
	+ '<a href="//other.com/d">Click</a>\n'
	+ '</b>';
const $ = cheerio.load(test);
let toFix = $('a[href^="//"]'); // a) <a>-tags with attr href starting with "//"
console.log(toFix.length + ' to fix');
toFix.each((idx, elem) => { // b) iterate over above tags one by one
	let $e = $(elem),  // c) elem is a bare node, $(elem) has method .attr()
		href = $e.attr('href');
	console.log(`fixing ${href}`);
	$e.attr('href', 'https:' + href);
});
```
output:
```
2 to fix
fixing //example.com/c
fixing //other.com/d
<b>
<a href="https://example.com/c">Click</a><br>
<a href="dont.fix.me/x">Click</a><br>
<a href="https://other.com/d">Click</a>
</b>
```
As you can see, using the API appropriately actually makes it more concise.
The three line comments a), b) and c) should make clear what's happening.

**Comment by rdm** [comment]
damn you jquery for promulgating an inconsistent data model... (and, ok, I should have double checked).

Anyways, good to know, thank you.

(But I think .each wasn't implemented a few days ago? I remember spending quite a bit of time trying to find some way of dealing with sequences because of that...)

**Comment by meisl** [comment]
Well, it was in 0.22.0 from August 2016. But I guess it was in there from the beginning, since it's pretty basic, and is rather trivial to implement:
```javascript
exports.each = function(fn) {
  var i = 0, len = this.length;
  while (i < len && fn.call(this[i], i, this[i]) !== false) ++i;
  return this;
};
```
(I think your for-loop was also broken)

**Comment by rdm** [comment]
Very odd - the docs I was reading specifically claimed that .each and .map were not yet implemented. (I suppose one possibility there was that google somehow landed me on a very stale copy of the docs and that I hadn't noticed the version mismatch.)

That said, I am curious about which part of my for-loop was broken (the termination condition or the advance-to-next clause)? I could see that it wasn't picking up the second [href], but I did not understand enough about the internals of the system to understand why that would be.

**Comment by meisl** [comment]
First off, let me say that I'm new to cheerio. too; evaluating it since a few days only (but I do know jquery a bit better).

As for the **documentation**: it seems there have been - and *right now are* - problems with that. It's not reflecting what's in 1.0.0-rc.2, both missing some things and also describing some things wrong.
However, that's not the case for .each. The snippet above is simply the source of 0.22.0 (and it's the same in master now).

Re the **for-loop**: since you asked I have dug a bit deeper - and now I'm puzzled! That is: I found other problems which I cannot explain. It might just be me, but maybe there is (additionally) something wrong with cheerio's .next()...

Well, I think we should first get a few things out of the way before we continue. Please let's be really precise with this, because it'll get tricky/strange enough afterwards:
1) `el.next('[href]')` in `for (let el= $('a[href]'); el.length; el= el.next('[href]'))` is superfluous, could just have been `el.next()`, according to the [jquery docs](https://api.jquery.com/next) - do you agree?
2) the actual output is NOT what you posted but rather
   ```
       2 'Click here' '//example.com'
       fixing //example.com/c
       <b><a href="https://example.com/c">Click here</a><br><a href="https://example.com/c">Click here</a></b>
   ```
   (you omitted the first line)
2) your expected output would have been
```
       2 'Click here' '//example.com/c'
       fixing //example.com/c
       1 'Click here' 'other.com/d'
       this looks ok: other.com/d
       <b><a href="https://example.com/c">Click here</a><br><a href="other.com/d">Click here</a></b>
   ```
   **notice the second `<a>` - it's *other.com/d*, NOT *example.com/c* as in the actual output**
3) having the *same* href-value in 2) - rather than different as in 3) - is due to calling .attr with *two args*

Again, please let's get these really clear first. Please confirm, or otherwise tell me what seems wrong.

**Comment by rdm** [comment]
1. Reading that jquery doc page, I see where I went wrong: .next() returns nothing if the next sibling doesn't match the selector.

2. Yes, it looks like I omitted the first line.

3. I don't get that from my originally posted code, but we might be working with different code at this point?

4. The point of the code was to change the href value - which seems to require calling .attr with two args. You posted a version different from my original which would work though.

I guess if I were using cheerio, I'd have to learn to ignore the cheerio docs and first read the jquery docs and then if cheerio's behavior were different read the cheerio docs (checking to make sure they were current docs) and only then if cheerio's behavior were still different from what I was reading would I have cause to report an issue...

Anyways.. I've closed this issue, because it looks like I had some serious understanding problems when I wrote it up.

(thanks)

**Comment by meisl** [comment]
Sure you closed this issue, since it turned out that cheerio's .attr is working as it should. That's totally fine.

1. Let's postpone this (.next()), it's actually at the core of the problem. And I now think I should have put that differently.
2. ok
3. I'm still referring to your original code. I was asking what you had *expected* - but didn't get, obviously. I figured you expected to go round the loop twice, the 2nd time landing in the `else` branch. Hence the additional lines
   ```
       1 'Click here' 'other.com/d'
       this looks ok: other.com/d
   ```
   Correct?
4. is about the other difference between *actual* output and *expected* output, namely in the 3rd line. There, in the 2nd `<a>` we got `href="https://example.com/c"` (*actual*) instead of `href="other.com/d"` (*expected*). But *that* we can explain (now) - it's because calling .attr(..) with 2 args before had already changed both `href` attributes.
   Right?

**Now, wrt. 3: how come the loop body seems to have been executed only once?**

---
p.s.: re the docs you're probably right...

**Comment by meisl** [comment]
p.p.s.: I'm trying this with both, cheerio and jsdom+jquery. For the latter:
```
npm install jsdom jquery
```
and then in `index.js`:
```javascript
    const test = // ...
    const { JSDOM } = require('jsdom'),
          dom = new JSDOM(test),
          $ = require('jquery')(dom.window);
    // ...
```
